artifactory/utils/wrappers/httpclient.go

```diff
@@ -0,0 +1,251 @@
+package wrappers
+
+import (
+	"github.com/jfrog/jfrog-client-go/artifactory/auth"
+	"github.com/jfrog/jfrog-client-go/artifactory/auth/cert"
+	artifactoryutils "github.com/jfrog/jfrog-client-go/artifactory/utils"
+	"github.com/jfrog/jfrog-client-go/errors/httperrors"
+	"github.com/jfrog/jfrog-client-go/httpclient"
+	"github.com/jfrog/jfrog-client-go/utils"
+	"github.com/jfrog/jfrog-client-go/utils/io/httputils"
+	"io"
+	"net/http"
+)
+
+type ArtifactoryHttpClientWrapper struct {
+	httpClient *httpclient.HttpClient
+	ArtDetails *auth.ArtifactoryDetails
+}
+
+func CreateArtifactoryHttpClient(artDetails *auth.ArtifactoryDetails, config artifactoryutils.Config) (*ArtifactoryHttpClientWrapper, error) {
+	if config.GetCertifactesPath() == "" {
+		return NewDefaultArtifactoryHttpClient(artDetails), nil
+	}
+
+	transport, err := cert.GetTransportWithLoadedCert(config.GetCertifactesPath())
+	if err != nil {
+		return nil, err
+	}
+	return NewArtifactoryHttpClient(artDetails, &http.Client{Transport: transport}), nil
+}
+
+func NewDefaultArtifactoryHttpClient(artDetails *auth.ArtifactoryDetails) *ArtifactoryHttpClientWrapper {
+	return NewArtifactoryHttpClient(artDetails, &http.Client{})
+}
+
+func NewArtifactoryHttpClient(artDetails *auth.ArtifactoryDetails, client *http.Client) *ArtifactoryHttpClientWrapper {
+	rtHttpClient := &ArtifactoryHttpClientWrapper{}
+	rtHttpClient.ArtDetails = artDetails
+	rtHttpClient.httpClient = httpclient.NewHttpClient(client)
+	return rtHttpClient
+}
+
+func (rtc *ArtifactoryHttpClientWrapper) SendGet(url string, followRedirect bool, httpClientsDetails *httputils.HttpClientDetails) (resp *http.Response, respBody []byte, redirectUrl string, err error) {
+	// A 'do while' loop that handles a token expiry if necessary
+	for isNewToken := true; isNewToken; {
```

Comment:
I suggest modifying all the loops in this files, so that they do not loop more than twice.
The current implementation may not cause a bug, but I guess they'd never need to run more than twice anyway, and it is best to avoid an endless loop (due to a future bug). Better to receive an error straight away than trying to figure our what causes the code to hang. 

And this raises another question - if I'm using the wrong SSH key, would the SSH login return an error or an 401 unauthorised? If the later is returned, then couldn't we end up with the an endless loop with the current implementation?

Reply:
I will limit the loop to run twice, and throw an error if a new token wasn't acquired by then.

When using the wrong SSH key (rsa key) the initial authentication will fail, so we will not get to this point. The program will exit on error in an earlier stage in this case.
Provided that the initial authenticating was successful, if the token expired the later is returned. The HandleTokenExpiry function is responsible for acquiring a new token, which will cause the loop to break.